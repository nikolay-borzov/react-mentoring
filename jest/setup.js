import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
