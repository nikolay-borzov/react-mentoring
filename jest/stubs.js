export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'http://via.placeholder.com/500x750',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre']
}

export const films = [film]
