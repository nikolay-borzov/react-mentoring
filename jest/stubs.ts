import { Film } from '../src/entities/film'

export const film: Film = {
  id: 1,
  title: 'Film title',
  tagline: 'Put that cookie down',
  vote_average: 7.5,
  vote_count: 667,
  release_date: '2018-05-04',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Once upon a time...',
  budget: 15000000,
  revenue: 59735548,
  genres: ['some genre', 'another genre']
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id: number) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
