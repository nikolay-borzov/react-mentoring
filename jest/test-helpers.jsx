import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
