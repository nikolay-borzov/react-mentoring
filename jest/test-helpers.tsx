import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
