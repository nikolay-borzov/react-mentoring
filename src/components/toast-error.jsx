import React from 'react'
import PropTypes from 'prop-types'

ToastError.propTypes = {
  message: PropTypes.string.isRequired,
  error: PropTypes.any
}

export function ToastError(props) {
  return (
    <div>
      {props.message}
      <br />
      {props.error.toString()}
    </div>
  )
}
