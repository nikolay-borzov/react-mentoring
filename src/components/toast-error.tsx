import React from 'react'

export interface ToastErrorProps {
  message: string
  error?: Error
}

export const ToastError: React.FunctionComponent<ToastErrorProps> = ({
  message,
  error
}) => (
  <div>
    {message}
    {error ? (
      <React.Fragment>
        <br />
        {error.toString()}
      </React.Fragment>
    ) : null}
  </div>
)
