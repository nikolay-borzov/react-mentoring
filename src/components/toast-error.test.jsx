import React from 'react'

import { itRendersCorrectly } from '../../jest/test-helpers'

import { ToastError } from './toast-error'

describe('ToastError component', () => {
  itRendersCorrectly(() => {
    const error = new Error('Some generic error')

    return <ToastError message="Error happened" error={error} />
  }, 'renders correctly with message and error details')

  itRendersCorrectly(
    () => <ToastError message="Error happened" />,
    'renders correctly with message'
  )
})
