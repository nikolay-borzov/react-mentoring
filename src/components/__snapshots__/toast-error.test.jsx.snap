// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastError component renders correctly with message 1`] = `
<div>
  Error happened
</div>
`;

exports[`ToastError component renders correctly with message and error details 1`] = `
<div>
  Error happened
  <React.Fragment>
    <br />
    Error: Some generic error
  </React.Fragment>
</div>
`;
